contracts/bonding/BondingVotes.sol

```diff
@@ -0,0 +1,534 @@
+        // delegator voting power calculated from events will only reflect their claimed stake without pending rewards.
+        if (previous.bondedAmount != current.bondedAmount) {
+            emit DelegatorVotesChanged(_account, previous.bondedAmount, current.bondedAmount);
+        }
+    }
+
+    /**
+     * @notice Returns whether an account already has any checkpoint.
+     * @dev This is meant to be called by a checkpoint initialization script once we deploy the checkpointing logic for
+     * the first time, so we can efficiently initialize the checkpoint state for all accounts in the system.
+     */
+    function hasCheckpoint(address _account) public view returns (bool) {
+        return bondingCheckpoints[_account].startRounds.length > 0;
+    }
+
+    /**
+     * @notice Called by the BondingManager when the total active stake changes.
+     * @dev This is called only from the {BondingManager-setCurrentRoundTotalActiveStake} function to set the total
+     * active stake in the current round.
+     * @param _totalStake From {BondingManager-currentRoundTotalActiveStake}
+     * @param _round The round for which the total active stake is valid. This is normally the current round.
+     */
+    function checkpointTotalActiveStake(uint256 _totalStake, uint256 _round) public virtual onlyBondingManager {
+        if (_round > clock()) {
+            revert FutureTotalStakeCheckpoint(_round, clock());
+        }
+
+        totalStakeCheckpoints.data[_round] = _totalStake;
+        totalStakeCheckpoints.rounds.pushSorted(_round);
+    }
+
+    // Historical stake access functions
+
+    /**
+     * @dev Gets the checkpointed total active stake at a given round.
+     * @param _round The round for which we want to get the total active stake.
+     */
+    function getTotalActiveStakeAt(uint256 _round) public view virtual returns (uint256) {
+        if (_round > clock() + 1) {
+            revert FutureLookup(_round, clock() + 1);
+        }
+
+        uint256 exactCheckpoint = totalStakeCheckpoints.data[_round];
+        if (exactCheckpoint > 0) {
+            return exactCheckpoint;
+        }
+
+        uint256[] storage initializedRounds = totalStakeCheckpoints.rounds;
+        uint256 upper = initializedRounds.findUpperBound(_round);
+        if (upper == 0) {
+            // Return a zero voting power supply for any round before the first checkpoint. This also happens if there
+            // are no checkpoints at all.
+            return 0;
+        } else if (upper < initializedRounds.length) {
+            // Use the checkpoint from the next initialized round, which got the next total active stake checkpointed.
+            uint256 nextInitedRound = initializedRounds[upper];
+            return totalStakeCheckpoints.data[nextInitedRound];
+        } else {
+            // Here the _round is after any initialized round, so grab its stake from nextRoundTotalActiveStake()
```

Comment:
Shouldn't ever get here right due to L325?

Reply:
We actually might in the following cases:
 - The current round has not been initialized yet, so instead of reverting or returning an inaccurate value until the round-initializer gets to it, we can just return `nextRoundTotalActiveStake`
 - Not only the current round, but a couple rounds have passed without having been initialized. All those rounds should use the `nextRoundTotalActiveStake` value as well to be internally consistent with the values we return for `getBondingCheckpointAt`. 
 
 This is related to that different behavior between individual & total checkpoints, explained in the other comment https://github.com/livepeer/protocol/pull/614#discussion_r1303341851
If we checkpointed the total active stake the "expensive way", every time `nextRound...` is changed, we could have the same logic for querying here: grab the lower bound checkpoint.
Since we checkpoint only on the next round instead, it means that the "uncheckpointed total stake value" stays in the `BondingManager` contract until another round is initialized. Which is why here we:
- Use "upper bound" instead of "lower bound" to find the stake of a previous round
- Use `nextRoundTotalActiveStake` if there are no rounds initialized after the searched round